src/lib/reasoners/intervals.ml

```diff
@@ -1200,6 +1200,150 @@ let pick ~is_max { ints; is_int; _ } =
        empty intervals. *)
     assert false
 
+(*****************)
+(*   Bit-vector  *)
+(*****************)
+
+(** Apply function [f] to the interval.
+
+    [f] *MUST* be monotone (either increasing or decreasing depending on the
+    value of the [increasing] labelled argument)
+
+    The [minfty] and [pinfty] optional arguments provide values for [map_bigint]
+    applied to [Minfty] and [Pinfty], i.e. the limits on negative and positive
+    infinities, respectively.
+
+    If no such value is provided, [Invalid_arg] is raised instead. *)
+let map_monotone ?minfty ?pinfty ~increasing f uints =
+  let get_infty = function
+    | Some infty -> infty
+    | None -> invalid_arg "Intervals.map"
+  in
+  let ints =
+    List.fold_left (fun l (lb, ub) ->
+        let f_lb =
+          match int_of_borne_inf lb with
+          | Large (lb, ex) -> Large (f lb, ex)
+          | Minfty -> get_infty minfty
+          | Strict _ | Pinfty -> assert false
+        in
+        let f_ub =
+          match int_of_borne_sup ub with
+          | Large (ub, ex) -> Large (f ub, ex)
+          | Pinfty -> get_infty pinfty
+          | Strict _ | Minfty -> assert false
+        in
+        let lb, ub = if increasing then f_lb, f_ub else f_ub, f_lb in
+        (lb, ub) :: l
+      ) [] uints.ints
+  in
+  let res = union_intervals { uints with ints } in
+  assert (res.ints != []);
```

Comment:
We need to normalize the result because `f` is not guaranteed to be *strictly* increasing/decreasing.

So for instance maybe we have `[10, 12] U [14, 20]` but `f(x) = 10` for any `x ≥ 10` so after the call we get `[10, 10] U [10, 10]` which is not properly normalized.

We need to support this because `shr` is not strictly increasing (all sufficiently small values map to `0`).

(interestingly enough, I believe this is the case where `union_intervals` currently has a subtle soundness bug (: )

Reply:
In English `increasing`means `strictement croissant`. So we should write [non-decreasing](https://fr.wiktionary.org/wiki/non-decreasing) :)

Reply:
`strictly increasing` is acceptable ;)